### PR TITLE
Migrate Caching things

### LINKED ISSUE #577: Migrate RCDeviceCache
Move the class method:
`+ (RCSubscriberAttribute *)newAttributeWithDictionary:(NSDictionary<NSString *, NSObject *> *)dictionary`
back to `SubscriberAttribute.swift` once this class is migrated.

This method exists because it cannot exist in Swift due to compiler bugs: https://bugs.swift.org/browse/SR-14847

Blocked by
- [x] #551 
- [ ] #578

### LINKED ISSUE #578: Migrate RCInMemoryCachedObject

### PR DESCRIPTION
**Pull request by taquitos**
Resolves #577
Resolves #578

### CODE REVIEW

File: PurchasesCoreSwift/Caching/DeviceCache.swift [hunk 1]
```diff
@@ -0,0 +1,497 @@
...
+
+    // Uses the accessQueue to synchronously write to the UserDefaults.
+    // Warning: do NOT nest calls to `writeCache`, it will deadlock.
+    // Only make calls to functions starting with "threadUnsafe" as those are guaranteed to not have any locking
+    // mechanisms.
+    private func writeCache(block: @escaping () -> Void) {
+        // .barrier is not needed here because we're using `.sync` instead of the normal .async multi-reader
+        // single-writer dispatch queue synchronization pattern. However, because we're using this as a way
+        // to control access to the datasource, passing `.barrier` flag is meant to help signal that intent.
+        accessQueue.sync(flags: .barrier) {
```
Comment (aboedo) [inline PurchasesCoreSwift/Caching/DeviceCache.swift:313]: would a serial queue work?

Comment (taquitos) [reply PurchasesCoreSwift/Caching/DeviceCache.swift:306]: Updated with a serial queue 😄

File: PurchasesCoreSwift/Caching/InMemoryCachedObject.swift [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+//
+//  Created by Joshua Liebowitz on 7/13/21.
+//  Copyright © 2021 Purchases. All rights reserved.
+//
+
+import Foundation
+
+open class InMemoryCachedObject<T> {
+
+    private let accessQueue = DispatchQueue(label: "InMemoryCachedObjectQueue", attributes: .concurrent)
```
[the remark this answers is not in the document]
Comment (taquitos) [reply PurchasesCoreSwift/Caching/InMemoryCachedObject.swift:13]: Unfortunately, unifying this queue doesn't make as much sense since this is used exclusively for maintaining exclusive access to a datasource. If we used a different queue for this, we might end up deadlocking if a work item in a global queue ends up doing a read/write to the datasource via `.sync`. This queue is a special purpose one, we really need to have very close control over it to avoid that sort of thing. Now that it is serial, and we're only dispatching `sync`, resource usage should be minimal.

File: PurchasesTests/Purchasing/PurchasesTests.swift [hunk 3]
```diff
@@ -12,8 +12,10 @@ import Purchases
 class PurchasesTests: XCTestCase {
 
     override func setUp() {
-        deviceCache = MockDeviceCache()
         userDefaults = UserDefaults(suiteName: "TestDefaults")
+        userDefaults.dictionaryRepresentation().keys.forEach { userDefaults.removeObject(forKey: $0) }
```
Comment (taquitos) [inline PurchasesTests/Purchasing/PurchasesTests.swift:16]: UserDefaults wasn't clean.

Comment (aboedo) [reply PurchasesTests/Purchasing/PurchasesTests.swift:16]: that must have been a fun one to figure out

File: PurchasesTests/Purchasing/PurchasesTests.swift [hunk 4]
```diff
@@ -12,8 +12,10 @@ import Purchases
...
+
+        deviceCache = MockDeviceCache(userDefaults: userDefaults)
```
Comment (aboedo) [inline PurchasesTests/Purchasing/PurchasesTests.swift:18]: I think this line is enough by itself, since we're removing the persistent domain on teardown, right?

Comment (taquitos) [reply PurchasesTests/Purchasing/PurchasesTests.swift:18]: You're 100% correct 😄

File: PurchasesCoreSwift/Caching/DeviceCache.swift [hunk 5]
```diff
@@ -0,0 +1,491 @@
...
+    private let assertionFunction: (String) -> Void
+
+    @objc convenience public init(userDefaults: UserDefaults = UserDefaults.standard) {
+        self.init(userDefaults: userDefaults, offeringsCachedObject: nil, notificationCenter: nil)
+    }
+
+    public init(userDefaults: UserDefaults = UserDefaults.standard,
+                offeringsCachedObject: InMemoryCachedObject<Offerings>? = InMemoryCachedObject(),
+                notificationCenter: NotificationCenter? = NotificationCenter.default,
+                assertionFunction: @escaping (String) -> Void = { assertionFailure($0) }) {
```
Comment (aboedo) [inline PurchasesCoreSwift/Caching/DeviceCache.swift:40]: why is this needed?

File: PurchasesCoreSwift/Caching/DeviceCache.swift [hunk 6]
```diff
@@ -0,0 +1,491 @@
...
+                                            name: UserDefaults.didChangeNotification,
+                                            object: self.userDefaults)
+    }
+
+    @objc private func handleUserDefaultsChanged(notification: Notification) {
+        guard let notificationObject = notification.object as? UserDefaults else {
+            return
+        }
+
+        if appUserIDHasBeenSet && notificationObject == self.userDefaults {
```
Comment (aboedo) [inline PurchasesCoreSwift/Caching/DeviceCache.swift:61]: nitpick: this bit might look a bit clearer as two distinct sets of conditions: 
1. it's the notification we care about
2. the app user id that we set has been removed

```swift
guard let notificationObject = notification.object as? UserDefaults, 
    notificationObject == self.userDefaults else { 
    return
}

if appUserIDHasBeenSet && threadUnsafeCachedAppUserID == nil {
    assertionFunction(...
}
```

File: PurchasesCoreSwift/Caching/DeviceCache.swift [hunk 7]
```diff
@@ -0,0 +1,491 @@
...
+            }
+        }
+    }
+
+    deinit {
+        notificationCenter.removeObserver(self)
+    }
+
+    // MARK: - appUserID
+    @objc(cacheAppUserID:) open func cache(appUserID: String) {
```
Comment (aboedo) [inline PurchasesCoreSwift/Caching/DeviceCache.swift:80]: these don't need to be `open` - when overriding for testing, we use `@testable import`

File: PurchasesCoreSwift/Caching/DeviceCache.swift [hunk 8]
```diff
@@ -0,0 +1,491 @@
...
+
+    @objc(subscriberAttributeWithKey:appUserID:)
+    open func subscriberAttribute(attributeKey: String, appUserID: String) -> SubscriberAttribute? {
+        return readCache { self.threadUnsafeStoredSubscriberAttributes(appUserID: appUserID)[attributeKey] }
+    }
+
+    // Threadsafe using accessQueue, however accessQueue is not reentrant. If you're calling this from somewhere
+    // that ends up in the accessQueue, it will deadlock. Instead, you'll want to call the version that isn't wrapped
+    // in a readCache{} block: threadUnsafeUnsyncedAttributesByKey
+    @objc open func unsyncedAttributesByKey(appUserID: String) -> [String: SubscriberAttribute] {
```
Comment (aboedo) [inline PurchasesCoreSwift/Caching/DeviceCache.swift:216]: I feel like the best way to go around these issues will be to refactor this into async calls and just not assume that the results will come in right away. 

Obviously that's a bigger refactor, but it's the approach that swift actors take and I think it'll simplify a lot of what we're doing.

File: PurchasesCoreSwift/Caching/DeviceCache.swift (hunk 1 above)
Comment (aboedo) [reply PurchasesCoreSwift/Caching/DeviceCache.swift:307]: if we're using a serial queue, `.barrier` no longer does anything, right? if not, we should remove the param and the comment

File: PurchasesCoreSwift/Caching/DeviceCache.swift [hunk 9]
```diff
@@ -0,0 +1,491 @@
...
+    public static func newAttribute(dictionary: [String: Any]) -> SubscriberAttribute {
+        // swiftlint:disable force_cast
+        let key = dictionary[SubscriberAttribute.keyKey] as! String
+        let value = dictionary[SubscriberAttribute.valueKey] as? String
+        let isSynced = (dictionary[SubscriberAttribute.isSyncedKey] as! NSNumber).boolValue
+        let setTime = dictionary[SubscriberAttribute.setTimeKey] as! Date
+        // swiftlint:enable force_cast
+
+        return SubscriberAttribute(withKey: key, value: value, isSynced: isSynced, setTime: setTime)
+    }
```
Comment (aboedo) [inline PurchasesCoreSwift/Caching/DeviceCache.swift:336]: why is this needed? 
if we need something for testing we should add it to the test targets instead

File: PurchasesCoreSwift/Caching/InMemoryCachedObject.swift [hunk 10]
```diff
@@ -0,0 +1,67 @@
...
+    }
+
+    open func clearCache() {
+        accessQueue.sync(flags: .barrier) {
+            self.lastUpdated = nil
+            self.cachedObject = nil
+        }
+    }
+
+    open func updateCacheTimestamp(date: Date) {
```
Comment (aboedo) [inline PurchasesCoreSwift/Caching/InMemoryCachedObject.swift:48]: why do these need to be `open`? if it's just for testing purposes, making them `internal` and using `@testable` imports should work

File: PurchasesCoreSwift/Caching/InMemoryCachedObject.swift [hunk 11]
```diff
@@ -0,0 +1,67 @@
...
+open class InMemoryCachedObject<T> {
+
+    private let accessQueue = DispatchQueue(label: "InMemoryCachedObjectQueue", attributes: .concurrent)
+    private var lastUpdated: Date?
+    private var cachedObject: T?
+    public var lastUpdatedAt: Date? {
+        accessQueue.sync {
+            return lastUpdated
+        }
+    }
```
Comment (aboedo) [inline PurchasesCoreSwift/Caching/InMemoryCachedObject.swift:20]: we should try to keep all the `public` stuff up top

File: PurchasesCoreSwift/Caching/InMemoryCachedObject.swift (hunk 2 above)
[the remark this answers is not in the document]
Comment (aboedo) [reply PurchasesCoreSwift/Caching/InMemoryCachedObject.swift:13]: same question here regarding `concurrent + .barrier` versus `serial`

File: PurchasesTests/Caching/DeviceCacheTests.swift [hunk 12]
```diff
@@ -198,31 +198,46 @@ class DeviceCacheTests: XCTestCase {
...
+        #if arch(arm64)
+        let assertionHappened = expectation(description: "Assertion happened")
+        self.deviceCache = DeviceCache(userDefaults: mockUserDefaults,
+                                       offeringsCachedObject: nil,
+                                       notificationCenter: mockNotificationCenter,
+                                       assertionFunction: { _ in assertionHappened.fulfill() })
+        #else
+        self.deviceCache = DeviceCache(userDefaults: mockUserDefaults,
+                                       offeringsCachedObject: nil,
+                                       notificationCenter: mockNotificationCenter)
```
Comment (aboedo) [inline PurchasesTests/Caching/DeviceCacheTests.swift:220]: why is this needed? does `toEventually` not do the trick here?

File: PurchasesCoreSwift/Caching/DeviceCache.swift (hunk 5 above)
Comment (taquitos) [reply PurchasesCoreSwift/Caching/DeviceCache.swift:40]: In order to test assertions in non-x86_64 platforms, we need to inject the assertion logic so that under normal circumstances, we we just `assertionFailure`, but during tests on ARM64, we just ensure this was called. You can check out the handling: https://github.com/RevenueCat/purchases-ios/pull/658/files#diff-f3a5243addd4ec9aa6a72782c2496bf77defcf852e8d6498ae3232bd88bbeddbR213

`assertionFailures` can't be tested anymore due to CPU architecture specifics 🙃

File: PurchasesCoreSwift/Caching/DeviceCache.swift (hunk 6 above)
Comment (taquitos) [reply PurchasesCoreSwift/Caching/DeviceCache.swift:61]: Ay, good catch.

File: PurchasesCoreSwift/Caching/DeviceCache.swift (hunk 8 above)
Comment (taquitos) [reply PurchasesCoreSwift/Caching/DeviceCache.swift:216]: Yeah, I was thinking exactly the same thing! Especially since the new `async functions` feature came out, it should simplify a lot of our logic too!

File: PurchasesCoreSwift/Caching/DeviceCache.swift (hunk 7 above)
Comment (taquitos) [reply PurchasesCoreSwift/Caching/DeviceCache.swift:80]: 🤦🤦🤦🤦🤦🤦🤦🤦🤦🤦🤦🤦🤦🤦🤦🤦🤦🤦🤦🤦🤦🤦
HOW DO I KEEP FORGETTING `@testable` ?

File: PurchasesCoreSwift/Caching/DeviceCache.swift (hunk 9 above)
Comment (taquitos) [reply PurchasesCoreSwift/Caching/DeviceCache.swift:336]: Obsolete comment, it's used internally 🙃  removed comment and changed access level to private.

File: PurchasesCoreSwift/Caching/InMemoryCachedObject.swift (hunk 10 above)
Comment (taquitos) [reply PurchasesCoreSwift/Caching/InMemoryCachedObject.swift:48]: Same as other open questions, I forgot 🤦

File: PurchasesTests/Caching/DeviceCacheTests.swift (hunk 12 above)
Comment (taquitos) [reply PurchasesTests/Caching/DeviceCacheTests.swift:220]: negative, can't test assertions in in `arm64` that way 😢

File: PurchasesCoreSwift/Caching/DeviceCache.swift (hunk 1 above)
Comment (taquitos) [reply PurchasesCoreSwift/Caching/DeviceCache.swift:307]: So, technically yeah, totally right and I agree. I'm thinking about future people looking at the code and changing it. The `.async(flags: .barrier)` pattern is pretty standard access control now, and I kinda think leaving this comment plus the `.barrier` flag to let people know "hey, this line of code is actually very important".  It's a signal to folks to reallllly consider what they are doing because it's pretty confusing.

So, I'm not sure what to do. If this were my own project, I'd leave this all here to remind my dumb future self, cause I'd look at the `.barrier` and go, wait, wut? and then think more.

Thoughts?

File: PurchasesCoreSwift/Caching/InMemoryCachedObject.swift (hunk 2 above)
[the remark this answers is not in the document]
Comment (taquitos) [reply PurchasesCoreSwift/Caching/InMemoryCachedObject.swift:13]: Same answer, `.barrier` is less common and gets people to think about what is going on more. It's a signal. I'm leaving these non-harmful things in to trigger people (probably myself) into thinking about what is happening more.

[the remark this answers is not in the document]
Comment (aboedo) [reply PurchasesCoreSwift/Caching/InMemoryCachedObject.swift:13]: 🤔 if the `.barrier` flag isn't doing anything other than signaling intent, perhaps a clearer way to signal the intent would be to create a wrapper with a clear name? 

i.e.: 
`accessQueue.sync(flags: .barrier) {` -> `accessQueue.executeByLockingDatasource {`?

[the remark this answers is not in the document]
Comment (aboedo) [reply PurchasesCoreSwift/Caching/InMemoryCachedObject.swift:13]: I'd still keep a comment to clarify what's going on

[the remark this answers is not in the document]
Comment (taquitos) [reply PurchasesCoreSwift/Caching/InMemoryCachedObject.swift:13]: Omg, great idea! Gonna do this.
